### PR TITLE
Bug 1599075 - Show Approve button when translation already exists.

### PR DESCRIPTION
**Pull request by adngdb**
This makes the actions behind the "Save" button of the Editor clearer: only two UI states where shown so far, but in reality three kinds of actions could happen. The "Approve" action was hidden, and this commit makes it explicit by changing the button label to show what is going to happen in the back-end. This also makes the front-end control the behavior instead of relying on the back-end, allowing us to make more focus backend views in a follow-up.

@mathjazz in order to avoid conflicts with my ActionLog PR, I want to do the back-end part later.

### CODE REVIEW

File: frontend/src/core/editor/components/EditorMainAction.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        if (translation === initialTranslation) {
+            sameTranslationAs = activeTranslation;
+        }
+        else if (history.translations.length) {
+            sameTranslationAs = history.translations.find(t => t.string === translation)
+        }
+    }
+
+    function approveTranslation() {
+        if (sameTranslationAs) {
```
Comment (mathjazz) [inline frontend/src/core/editor/components/EditorMainAction.js]: Is this check necessary? Seems like `approveTranslation()` can only be called if the condition is true.

File: frontend/src/core/editor/components/EditorMainAction.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            updateTranslationStatus(sameTranslationAs.pk, 'approve');
+        }
+    }
+
+    if (isTranslator && sameTranslationAs && !sameTranslationAs.approved) {
+        // Approve button, will approve the translation.
+        return <Localized
+            id="editor-EditorMenu--button-approve"
```
Comment (mathjazz) [inline frontend/src/core/editor/components/EditorMainAction.js]: Please also update the FTL file accordingly.

File: frontend/src/core/editor/components/EditorMainAction.test.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            translation: 'hello, world!',
+        }
+        const wrapper = shallow(<EditorMainAction { ...params } />);
+
+        expect(wrapper.find('.action-save')).toHaveLength(1);
+        expect(wrapper.find('.action-suggest')).toHaveLength(0);
+        expect(wrapper.find('.action-approve')).toHaveLength(0);
+    });
+
+    it('renders the Save button when user does not have permission', () => {
```
Comment (mathjazz) [inline frontend/src/core/editor/components/EditorMainAction.test.js]: I think we have a bug (here and in prod): we should show the "Suggest" button for non-privileged users.

https://mozilla-l10n.github.io/localizer-documentation/tools/pontoon/translate.html#submit-a-suggestion

File: frontend/src/core/editor/components/EditorMainAction.test.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        expect(wrapper.find('.action-suggest')).toHaveLength(0);
+        expect(wrapper.find('.action-approve')).toHaveLength(0);
+    });
+
+    it('renders the Save button when same translation exists but is approved', () => {
+        const params = {
+            initialTranslation: 'something',
+            isTranslator: true,
+            history: HISTORY,
+            translation: 'hello, world!',
```
Comment (mathjazz) [inline frontend/src/core/editor/components/EditorMainAction.test.js]: Translations are not the same, which is what the test title says.

File: frontend/src/core/editor/components/EditorMainAction.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        initialTranslation,
+        isTranslator,
+        forceSuggestions,
+        history,
+        translation,
+        sendTranslation,
+        updateTranslationStatus,
+    } = props;
+
+    let sameTranslationAs = null;
```
Comment (mathjazz) [inline frontend/src/core/editor/components/EditorMainAction.js]: Maybe `existingTranslation`?

File: frontend/src/core/editor/components/EditorMainAction.js (hunk 1 above)
Comment (adngdb) [reply frontend/src/core/editor/components/EditorMainAction.js]: It's needed for Flow. :sweat_smile:

File: frontend/src/core/editor/components/EditorMainAction.test.js (hunk 4 above)
Comment (adngdb) [reply frontend/src/core/editor/components/EditorMainAction.test.js]: It is the same as one that exists in the history. :)

Comment (mathjazz) [reply frontend/src/core/editor/components/EditorMainAction.test.js]: Ah... Maybe add "exists **in history**" to make it more obvious?

File: frontend/src/core/editor/components/EditorMainAction.test.js (hunk 3 above)
Comment (adngdb) [reply frontend/src/core/editor/components/EditorMainAction.test.js]: Nice catch! Thank you unit tests! :D

**Comment by mathjazz** [comment]
Excellent job! There's just one thing we should fix - if a check fails when submitting a translation, we show the green "Save Anyway" button instead of the blue "Suggest Anyway" button to non-privileged users.

**Comment by adngdb** [comment]
I realized while working on the "Save Anyway" feature that I forgot to code the correct behavior for the "Enter" shortcut. It's a big problem as it will break the feature when we remove the back-end equivalent feature.

**Comment by adngdb** [comment]
@mathjazz it turned out to require a bunch more work, but hopefully everything works now! (Main editor action button, Enter shortcut, and FailedChecks. )

**Comment by adngdb** [comment]
@mathjazz done! :-)

**Pull request merged**